File: JVM/jnetcall-java-common/src/main/java/jnetcall/java/impl/io/net/UdpTransport.java
package jnetcall.java.impl.io.net;

import java.io.IOException;
import java.net.SocketAddress;
import java.nio.ByteBuffer;
import java.nio.channels.DatagramChannel;

import jnetcall.java.api.enc.IEncoding;
import jnetcall.java.api.io.IPullTransport;
import jnetcall.java.api.io.ISendTransport;

public final class UdpTransport implements ISendTransport, IPullTransport, AutoCloseable {

    private final IEncoding<byte[]> _encoding;
    private final DatagramChannel _receiver;
    private final DatagramChannel _sender;

    public UdpTransport(IEncoding<byte[]> encoding,
                         String hostIn, int portIn, String hostOut, int portOut) {
        this(encoding, NetworkTools.toEndPoint(hostIn, portIn),
                NetworkTools.toEndPoint(hostOut, portOut));
    }

    private UdpTransport(IEncoding<byte[]> encoding,
                          SocketAddress endPointIn, SocketAddress endPointOut) {
        _encoding = encoding;
        try {
            _receiver = DatagramChannel.open().bind(endPointIn);
            _sender = DatagramChannel.open();
            _sender.connect(endPointOut);
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
    }

    @Override
    public <T> T pull(Class<T> clazz) {
        try {
            ByteBuffer buff = ByteBuffer.allocate(65535);
            _receiver.receive(buff);
            return _encoding.decode(buff.array(), clazz);
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }

    @Override
    public <T> void send(T payload) {
        try {
            byte[] bytes = _encoding.encode(payload);
            ByteBuffer buff = ByteBuffer.wrap(bytes);
            _sender.write(buff);
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }

    @Override
    public void close() throws Exception {
        _encoding.close();
        _sender.close();
        _receiver.close();
    }
}

File: NET/JNetCall.Sharp.Common/Impl/IO/Net/UdpTransport.cs
using System.Net;
using System.Net.Sockets;
using JNetCall.Sharp.API.Enc;
using JNetCall.Sharp.API.IO;

namespace JNetCall.Sharp.Impl.IO.Net
{
    public sealed class UdpTransport : IPullTransport
    {
        private readonly IEncoding<byte[]> _encoding;
        private readonly UdpClient _receiver;
        private readonly UdpClient _sender;

        public UdpTransport(IEncoding<byte[]> encoding,
            string hostIn, int portIn, string hostOut, int portOut)
            : this(encoding, NetworkTools.ToEndPoint(hostIn, portIn),
                NetworkTools.ToEndPoint(hostOut, portOut))
        {
        }

        private UdpTransport(IEncoding<byte[]> encoding,
            IPEndPoint endPointIn, IPEndPoint endPointOut)
        {
            _encoding = encoding;
            _receiver = new UdpClient(endPointIn);
            _sender = new UdpClient();
            _sender.Connect(endPointOut);
        }

        public void Send<T>(T payload)
        {
            var bytes = _encoding.Encode(payload);
            _sender.Send(bytes, bytes.Length);
        }

        private IPEndPoint _remote;

        public T Pull<T>()
        {
            var bytes = _receiver.Receive(ref _remote);
            return _encoding.Decode<T>(bytes);
        }

        public void Dispose()
        {
            _encoding.Dispose();
            _sender.Close();
            _sender.Dispose();
            _receiver.Close();
            _receiver.Dispose();
        }
    }
}